### DISCUSSION TITLE
Reload main script from submodule

### DISCUSSION THREAD

**Discussion by schoeller** [CEXT-Dan/PyRx#142] [Q&A] [answered]
Dear all,

I have the following directory structure for a script collection:

```
/commands.py
=>PyRxCmd_bct_reloadcommands()
calling 
helper.reload_commands()
from
/helper/helper.py
=>reload_commands()
```

Within commands.py used modules are reloaded once the script is reloaded through _PYRELOAD as such:

```python
def OnPyReload() -> None:
    try:
        import importlib
        importlib.reload(helper)
        print("\nReloading all modules")
    except Exception as err:
        print(err)
```

As time permits I am adding command line functions in `commands.py` which point to function/modules in subfolders. 
What I am trying to achieve is a command to reload also `commands.py` in itself via `_BCT_RELOADCOMMANDS`. I have fumbled with `importlib`, but no resolution.
Ideas welcome.

Cheers Seb

**Comment by CEXT-Dan** [comment]
I haven’t made any large applications using Python yet, so I’m not sure of best practices with regards for code layout and reloading.
Maybe use a package with __init__.py? I still have little knowledge of python

**Reply by schoeller** [reply]
How would I call alternatively _PYRELOAD from a script?

**Reply by CEXT-Dan** [reply, marked as answer]
There’s some utilities in Ap.Application, maybe one of these will work for you

```Python
Ap.Application.loadPythonModule(path)
Ap.Application.reloadPythonModule(path)
Ap.Application.getLoadedModules()
Ap.Application.getLoadedModuleNames()
```

```Python 
def PyRxCmd_doit():
    try:
        for modulePath in Ap.Application.getLoadedModules():
            Ap.Application.reloadPythonModule(modulePath)
    except Exception as err:
        traceback.print_exception(err)
```

**Reply by schoeller** [reply]
For some reason I have to call it twice for changes to appear. I have embedded this with a print statement right afterwards. When changing the print statement I have to call reload twice for file changes to appear.
